Clarify Yul scoping rules. (#11726)
ethereum/solidity: docs/yul.rst

introduce new identifiers into these scopes.

Identifiers are visible in

> the block they are defined in (including all sub-nodes and sub-blocks),
> functions also before their declaration but variables only starting from the statement after
> the ``VariableDeclaration``.

Reviewer: ```suggestion
the block they are defined in (including all sub-nodes and sub-blocks).
Variables are only visible from the statement after their declaration (using ``let varName := ...``), whereas functions are also visible before their declaration.
```

Author: I thought about that, but if we add a full stop there, the first sentence contracdicts the second.

Reviewer: I'm okay with keep it together, but the line

> functions also before their declaration but variables

doesn't sound right.

> This means that variables can only be referenced after their declaration. In particular,
> variables cannot be referenced in the right hand side of their own variable
> declaration.

Reviewer: ```suggestion
In particular,
variables cannot be referenced in the right hand side of their own variable
declaration.
```

This means a for-loop of the form ``for { I... } C { P... } { B... }`` is equivalent
to ``{ I... for {} C { P... } { B... } }``.

The parameters and return parameters of functions are visible in the
function body and their names have to be distinct.

Inside functions, it is not possible to reference a variable that was declared
outside of that function.

> 

Reviewer: ```suggestion
```

Reviewer: This is what I meant in https://github.com/ethereum/solidity/pull/11726/files#r680945865

Author: Hm, I think shadowing is important enough so that it gets its own paragraph. And the one before is still about functions and visibilty.

Reviewer: Isn't shadowing only relevant for functions?

Author: No, also inside blocks:
```
{
let x
  {
    let x
  }
}
```

Shadowing is disallowed, i.e. you cannot declare an identifier at a point
where another identifier with the same name is also visible, even if it is

> not possible to reference it because it was declared outside the current function.

Reviewer: I think combining these two paragraphs would be good.

Author: Sorry - which two paragraphs?
